Point [SOURCE] documents to github

dmatos2012 opened the pull request:
## PR Summary
Addresses #20557 to point the [SOURCE] in docs point to github instead. A proper build config to build the old way is still missing, but perhaps that can be part of separate PR.  Also, this impl was copied from [scipy conf.py](https://github.com/scipy/scipy/blob/master/doc/source/conf.py#L426-L502), but modified slightly to work here instead. 

PS: Building old way is just adding back the `sphinx.ext.viewcode`, but I was not sure how to add it as part of a `make O=-Dmyarg=1 html`. 

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->

- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [ ] New features are documented, with examples if plot related.
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of master, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

jklymak commented:
This looks good.  I'm surprised that there ins't a built in sphinx option for this though....

QuLogic commented on doc/conf.py:
```diff
@@ -530,3 +529,80 @@ def setup(app):
+# -----------------------------------------------------------------------------
+link_github = True
+# You can add viewcode extension to build old way with link_github = False
+# extensions.append('sphinx.ext.viewcode')
```
You should have the `else` do that for you.

QuLogic commented on doc/conf.py:
```diff
@@ -530,3 +529,80 @@ def setup(app):
+    import re
+    import inspect
+    from os.path import relpath, dirname
+    for name in ['sphinx.ext.linkcode', 'linkcode', 'numpydoc.linkcode']:
```
`sphinx.ext.linkcode` was added in Sphinx 1.2, and we require 1.8.1, so I don't know why you need all the fallbacks.

dmatos2012 commented:
Sorry for the delay, was caught up with the non-digital world. I have addressed the comments by @QuLogic. I think now it should be fine hopefully, otherwise I can make changes np.

jklymak commented:
Many of these links don't seem to go to the right place, though _some_ of them do...

dmatos2012 commented:
Are these the ones that are deprecated? While I tested it, I saw those would go to the deprecation github page. Aside from that, also [add_callback](https://github.com/matplotlib/matplotlib/blob/v3.4.2/lib/matplotlib/artist.py#L365-L392) confuses the function with the docstring `add_callback`. Any other ones? In that case, I think its the `inspect` functions themselves, but yeah not sure.

jklymak commented:
I only did a spot check, but starting from https://62590-1385122-gh.circle-artifacts.com/0/doc/build/html/api/axes_api.html `barh` works, but `bar` does not.  `fill_between` and `fill_betweenx` do not work,  for me at least (Safari). i.e. `fill_between` links to:  https://github.com/matplotlib/matplotlib/blob/300e9cdb/lib/matplotlib/__init__.py#L5259-L5263

dmatos2012 commented:
So now those ones you spotted are fixed. I also checked around ~20 more of those, and they look ok as well.

jklymak commented:
Great.  Do you know why they were different?

dmatos2012 commented:
I noticed `inspect.getsourcefile(obj)` was returning the filename to be  `lib/matplotlib/__init__.py` instead of the python file name in which it was defined. Thus, I added  `fn.endswith('__init__.py') `to get the filename like shown below and then it shows the proper filename. 
```python
if not fn or fn.endswith('__init__.py'):
    fn = inspect.getsourcefile(sys.modules[obj.__module__])
```

QuLogic commented on doc/conf.py:
```diff
@@ -534,3 +533,75 @@ def setup(app):
+        startdir = os.path.abspath(os.path.join(dirname(matplotlib.__file__), '..'))
+        fn = relpath(fn, start=startdir).replace(os.path.sep, '/')
+
+        if fn.startswith('matplotlib/') or fn.startswith('mpl_toolkits/'):
+            m = re.match(r'^.*post[0-9]+\+\w([a-z0-9]+).\w+$', matplotlib.__version__)
+            if m:
+                return "https://github.com/matplotlib/matplotlib/blob/%s/lib/%s%s" % (
+                        m.group(1), fn, linespec)
+            else:
+                return "https://github.com/matplotlib/matplotlib/blob/v%s/lib/%s%s" % (
+                        matplotlib.__version__, fn, linespec)
+        else:
+            return None
```
```suggestion
        if not fn.startswith(('matplotlib/', 'mpl_toolkits/')):
            return None

        m = re.match(r'^.*post[0-9]+\+\w([a-z0-9]+).\w+$', matplotlib.__version__)
        if m:
            return "https://github.com/matplotlib/matplotlib/blob/%s/lib/%s%s" % (
                    m.group(1), fn, linespec)
        else:
            return "https://github.com/matplotlib/matplotlib/blob/v%s/lib/%s%s" % (
                    matplotlib.__version__, fn, linespec)
```

QuLogic commented on doc/conf.py:
```diff
@@ -534,3 +533,75 @@ def setup(app):
+        else:
+            linespec = ""
+
+        startdir = os.path.abspath(os.path.join(dirname(matplotlib.__file__), '..'))
+        fn = relpath(fn, start=startdir).replace(os.path.sep, '/')
```
```suggestion
        startdir = Path(matplotlib.__file__).parent.parent
        fn = os.path.relpath(fn, start=startdir).replace(os.path.sep, '/')
```

dmatos2012 commented:
@QuLogic I added some extra `Exceptions`  and added the rest of your suggestions.
